katcp/server.py

```diff
@@ -2318,6 +2338,99 @@ def inform_callback(sensor, reading):
         yield gen.moment
         raise gen.Return(req.make_reply("ok", name, strategy, *params))
 
+    @gen.coroutine
+    def _handle_bulk_set_sensor_sampling(self, req, msg):
+        if not msg.arguments:
+            raise FailReply("No sensor name given.")
+
+        # Arguments start with one or more sensor names.  Check the first one now.
+        # A sensor object is required for testing the sampling strategies, so
+        # the first one is as good as any.
+        name = msg.arguments[0]
+        if name not in self._sensors:
+            raise FailReply("Unknown sensor name: %s." % name)
+        sensor = self._sensors[name]
+
+        # The client connection that is not specific to this request context
+        client = req.client_connection
+        katcp_version = self.PROTOCOL_INFO.major
+        format_inform = (format_inform_v5
+                         if katcp_version >= SEC_TS_KATCP_MAJOR
+                         else format_inform_v4)
+
+        def inform_callback(sensor, reading):
+            """Inform callback for sensor strategy."""
+            timestamp, status, value = reading
+            cb_msg = format_inform(sensor, timestamp, status, value)
+            client.inform(cb_msg)
+
+        # Parse arguments from the back, trying all possible strategies until
+        # a valid one is found.  The number of arguments to define a strategy
+        # is one for the name, plus the number of params expected for that type
+        # of strategy.
+        # E.g., for 'none', only 1 argument.  For 'event-rate 5 60', 3 arguments.
+        requested_strategy = None
+        min_args = 1
+        max_args = 1 + max(SampleStrategy.SAMPLING_NUM_PARAMS_LOOKUP.values())
+        for num_strategy_args in range(min_args, max_args + 1):
+            args = msg.arguments[-num_strategy_args:]
+            try:
+                strategy = args[0]
+                params = args[1:]
+                if self.PROTOCOL_INFO.strategy_allowed(strategy):
+                    if katcp_version < SEC_TS_KATCP_MAJOR and strategy == 'period':
+                        # Slightly nasty hack, but since period is the only v4 strategy
+                        # involving timestamps it's not _too_ nasty :)
+                        params = [float(params[0]) * MS_TO_SEC_FAC] + params[1:]
+                    requested_strategy = SampleStrategy.get_strategy(
+                        strategy, inform_callback, sensor, *params, ioloop=self.ioloop)
+                    if requested_strategy:
+                        break
+            except Exception:
+                pass
+
+        if not requested_strategy:
+            raise FailReply("No valid strategy provided: %s." % msg.arguments)
+
+        # with known strategy, remaining arguments are sensor names
```

Comment:
May be change comment for clarity "preceding arguments are sensor names" instead of "remaining"
(or it could be a slightly incorrect Afrikaans definition of "remaining")

Reply:
Thanks.  Using "preceding" would be better.